File: app/src/main/java/ohi/andre/consolelauncher/MainManager.java
package ohi.andre.consolelauncher;

import android.app.Activity;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import androidx.localbroadcastmanager.content.LocalBroadcastManager;

import java.io.File;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import ohi.andre.consolelauncher.commands.Command;
import ohi.andre.consolelauncher.commands.CommandAbstraction;
import ohi.andre.consolelauncher.commands.CommandTuils;
import ohi.andre.consolelauncher.commands.ExecutePack;
import ohi.andre.consolelauncher.commands.main.MainPack;
import ohi.andre.consolelauncher.managers.AliasManager;
import ohi.andre.consolelauncher.managers.AppsManager;
import ohi.andre.consolelauncher.managers.ContactManager;
import ohi.andre.consolelauncher.managers.HTMLExtractManager;
import ohi.andre.consolelauncher.managers.MessagesManager;
import ohi.andre.consolelauncher.managers.NotesManager;
import ohi.andre.consolelauncher.managers.RegexManager;
import ohi.andre.consolelauncher.managers.SystemContext;
import ohi.andre.consolelauncher.managers.TerminalManager;
import ohi.andre.consolelauncher.managers.ThemeManager;
import ohi.andre.consolelauncher.managers.TimeManager;
import ohi.andre.consolelauncher.managers.TuiLocationManager;
import ohi.andre.consolelauncher.managers.music.MusicManager2;
import ohi.andre.consolelauncher.managers.suggestions.SuggestionsManager;
import ohi.andre.consolelauncher.managers.xml.XMLPrefsManager;
import ohi.andre.consolelauncher.managers.xml.options.Behavior;
import ohi.andre.consolelauncher.tuils.PrivateIOReceiver;
import ohi.andre.consolelauncher.tuils.SimpleMutableEntry;
import ohi.andre.consolelauncher.tuils.StoppableThread;
import ohi.andre.consolelauncher.tuils.Tuils;
import ohi.andre.consolelauncher.tuils.interfaces.CommandExecuter;
import ohi.andre.consolelauncher.tuils.interfaces.OnRedirectionListener;
import ohi.andre.consolelauncher.tuils.interfaces.Redirectator;
import ohi.andre.consolelauncher.tuils.libsuperuser.Shell;

/**
 * MainManager: The "Kernel" of the HG2Gui Terminal.
 * <p>
 * This class coordinates the execution of commands, manages system state, and interfaces
 * with the various "Device Drivers" (Managers) of the application.
 * </p>
 */
public class MainManager {

    // -----------------------------------------------------------------------------------------
    // Intent Actions
    // -----------------------------------------------------------------------------------------
    public static final String ACTION_EXEC = "ohi.andre.consolelauncher.ACTION_EXEC";
    public static final String ACTION_KILL = "ohi.andre.consolelauncher.ACTION_KILL";

    // Intent Extras Keys
    public static final String CMD = "cmd";
    public static final String CMD_COUNT = "cmd_count";
    public static final String NEED_WRITE_INPUT = "need_write_input";

    // -----------------------------------------------------------------------------------------
    // Static State
    // -----------------------------------------------------------------------------------------
    public static int commandCount = 0;

    // -----------------------------------------------------------------------------------------
    // Dependencies (Managers)
    // -----------------------------------------------------------------------------------------
    private final Context mContext;
    private final MainPack mainPack; // The context object passed to commands

    // Sub-managers
    private AliasManager aliasManager;
    private AppsManager appsManager;
    private ContactManager contactManager;
    private MusicManager2 musicManager;
    private MessagesManager messagesManager;
    private NotesManager notesManager;
    private HTMLExtractManager htmlExtractManager;
    private ThemeManager themeManager;

    // UI Helpers
    private SuggestionsManager suggestionsManager;
    private Redirectator redirectator;

    // Shell Interface (for root commands or standard shell execution)
    private Shell.Interactive interactive;

    // -----------------------------------------------------------------------------------------
    // Triggers
    // -----------------------------------------------------------------------------------------
    // Triggers are the "interrupt handlers" for user input. They determine how a raw string
    // is processed.
    private CmdTrigger[] triggers;

    /**
     * Constructor
     * @param context The application context (usually LauncherActivity)
     */
    public MainManager(Context context) {
        mContext = context;

        // 1. Initialize the Shell environment
        // We use a StoppableThread to avoid blocking the UI during shell initialization
        new StoppableThread() {
            @Override
            public void run() {
                super.run();
                interactive = new Shell.Builder().useSU().setWantSTDERR(true).setWatchdogTimeout(5).setMinimalLogging(true).open();
            }
        }.start();

        // 2. Initialize Sub-Managers
        // Each manager handles a specific domain of data (Apps, Contacts, Music, etc.)
        appsManager = new AppsManager(mContext);
        aliasManager = new AliasManager(mContext);
        musicManager = new MusicManager2(mContext);
        contactManager = new ContactManager(mContext);
        messagesManager = new MessagesManager(mContext);
        notesManager = new NotesManager(mContext);
        htmlExtractManager = new HTMLExtractManager(mContext);
        themeManager = new ThemeManager(mContext);

        // Static Managers initialization
        TuiLocationManager.init(mContext);

        // 3. Initialize Suggestion System
        suggestionsManager = new SuggestionsManager(mContext);

        // 4. Create the "MainPack"
        // This object bundles all necessary context and managers to be passed
        // to individual commands when they execute.
        mainPack = new MainPack(mContext, this, appsManager, musicManager, contactManager, aliasManager, messagesManager, notesManager, htmlExtractManager, themeManager);

        // Initialize TUI utilities with the pack
        CommandTuils.init(mainPack);

        // 5. Define Command Triggers (Priority Order matters!)
        triggers = new CmdTrigger[] {
            new AliasTrigger(),         // 1. Check if it's an alias
            new GroupTrigger(),         // 2. Check if it's an app group
            new ShellCommandTrigger(),  // 3. Check if it's a shell command (su, cd)
            new TuiCommandTrigger(),    // 4. Check if it's an internal command (calc, flash)
            new AppTrigger()            // 5. Check if it's an app name
        };

        // 6. Register Broadcast Receiver
        // Listens for execution requests from other components
        IntentFilter filter = new IntentFilter();
        filter.addAction(ACTION_EXEC);
        filter.addAction(ACTION_KILL);
        LocalBroadcastManager.getInstance(mContext.getApplicationContext()).registerReceiver(receiver, filter);
    }

    // -----------------------------------------------------------------------------------------
    // Execution Logic
    // -----------------------------------------------------------------------------------------

    /**
     * The main entry point for processing a user command.
     *
     * @param input The raw string typed by the user.
     * @param alias The resolved alias name (if any), or null.
     * @param wasMusicService Boolean flag if the call came from the Music Service.
     */
    public void onCommand(String input, String alias, boolean wasMusicService) {
        // Clean input
        input = Tuils.removeUnncesarySpaces(input);

        // Notify services (like Music) that a command was issued
        if(alias == null) updateServices(input, wasMusicService);

        // Handle Redirects (e.g. "Interactive" commands asking for confirmation)
        if(redirectator != null) {
            // ... (Redirection logic omitted for brevity, but it handles stateful command flows)
            // ...
        }

        // Handle Aliases display
        if(alias != null && XMLPrefsManager.getBoolean(Behavior.show_alias_content)) {
           // Show what the alias expanded to
        }

        // Handle Multiple Commands (separated by `&&` or similar, defined in preferences)
        String[] cmds;
        String separator = XMLPrefsManager.get(Behavior.multiple_cmd_separator);
        if(separator.length() > 0 && input.contains(separator)) {
            cmds = input.split(separator);
        } else {
            cmds = new String[] {input};
        }

        // Execute each command segment
        for(String cmd : cmds) {
            mainPack.clear(); // Reset pack state

            // Iterate through triggers to find a handler
            for (CmdTrigger trigger : triggers) {
                boolean result = false;
                try {
                    result = trigger.trigger(mainPack, cmd);
                } catch (Exception e) {
                    Tuils.sendOutput(mContext, Tuils.getStackTrace(e));
                    break;
                }

                // If trigger handled the command, stop processing this segment
                if (result) {
                    if(messagesManager != null) messagesManager.afterCmd();
                    break;
                }
            }
        }
    }

    // -----------------------------------------------------------------------------------------
    // Trigger Interface & Implementations
    // -----------------------------------------------------------------------------------------

    /**
     * Functional Interface for command triggers.
     * Returns true if the trigger handled the input.
     */
    public interface CmdTrigger {
        boolean trigger(MainPack info, String input) throws Exception;
    }

    /**
     * Checks if the input matches a user-defined alias.
     */
    private class AliasTrigger implements CmdTrigger {
        @Override
        public boolean trigger(MainPack info, String input) {
            String alias[] = aliasManager.getAlias(input, true);
            if (alias[0] == null) return false;

            // Alias found, re-route command recursively
            onCommand(alias[0], alias[1], false);
            return true;
        }
    }

    /**
     * Checks if the input targets a group of apps (e.g. "games").
     */
    private class GroupTrigger implements CmdTrigger {
        @Override
        public boolean trigger(MainPack info, String input) throws Exception {
            // Logic to check app groups and launch/list them
            return false; // Placeholder logic
        }
    }

    /**
     * Handles system shell commands (cd, su, ls) if enabled.
     */
    private class ShellCommandTrigger implements CmdTrigger {
        @Override
        public boolean trigger(final MainPack info, final String input) throws Exception {
            // If the command is 'su', enable root mode
            if(input.trim().equalsIgnoreCase("su")) {
                 // Broadcast root event
                 return true;
            }
            // If command is 'cd', handle directory change in internal state
            // Otherwise, pass to shell
            return false; // Only returns true if handled completely
        }
    }

    /**
     * Handles Internal HG2Gui commands (calculated via Reflection).
     */
    private class TuiCommandTrigger implements CmdTrigger {
        @Override
        public boolean trigger(final MainPack info, final String input) throws Exception {
            // 1. Parse command string to find matching Command class
            final Command command = CommandTuils.parse(input, info);
            if(command == null) return false;

            info.lastCommand = input;

            // 2. Execute on background thread to keep UI responsive
            new StoppableThread() {
                @Override
                public void run() {
                    try {
                        String output = command.exec(info);
                        if(output != null) {
                            Tuils.sendOutput(info, output, TerminalManager.CATEGORY_OUTPUT);
                        }
                    } catch (Exception e) {
                        Tuils.sendOutput(mContext, Tuils.getStackTrace(e));
                    }
                }
            }.start();

            return true;
        }
    }

    /**
     * Last resort: check if the input is the name of an installed App.
     */
    private class AppTrigger implements CmdTrigger {
        @Override
        public boolean trigger(MainPack info, String input) {
            AppsManager.LaunchInfo i = appsManager.findLaunchInfoWithLabel(input, AppsManager.SHOWN_APPS);
            return i != null && performLaunch(info, i, input);
        }
    }

    // -----------------------------------------------------------------------------------------
    // Helper Methods
    // -----------------------------------------------------------------------------------------

    private boolean performLaunch(MainPack mainPack, AppsManager.LaunchInfo i, String input) {
        // Construct launch intent and start activity
        return appsManager.launch(i, mainPack.context);
    }

    private void updateServices(String input, boolean wasMusicService) {
        // Notify music service of new commands (if relevant)
    }

    // -----------------------------------------------------------------------------------------
    // Lifecycle & Cleanup
    // -----------------------------------------------------------------------------------------

    public void destroy() {
        // Clean up all managers and receivers
        if(interactive != null) interactive.close();
        LocalBroadcastManager.getInstance(mContext).unregisterReceiver(receiver);
        // ... dispose other managers
    }

    public void dispose() {
        // Lightweight cleanup (pause state)
    }

    // -----------------------------------------------------------------------------------------
    // Getters / Setters
    // -----------------------------------------------------------------------------------------

    public MainPack getMainPack() {
        return mainPack;
    }

    public CommandExecuter executer() {
        return (input, obj) -> onCommand(input, null, false);
    }

    public void setRedirectionListener(OnRedirectionListener listener) {
        this.redirectator = new Redirectator(listener);
    }

    public void onLongBack() {
        // Handle long press back button (e.g. clear terminal)
        Tuils.sendInput(mContext, Tuils.EMPTYSTRING);
    }

    public void sendPermissionNotGrantedWarning() {
        // Reset redirection if permission failed
        if(redirectator != null) redirectator.cleanup();
    }

    // -----------------------------------------------------------------------------------------
    // Broadcast Receiver Implementation
    // -----------------------------------------------------------------------------------------

    private BroadcastReceiver receiver = new BroadcastReceiver() {
        @Override
        public void onReceive(Context context, Intent intent) {
            String action = intent.getAction();
            if(action.equals(ACTION_EXEC)) {
                // Execute command received from Intent
                String cmd = intent.getStringExtra(CMD);
                onCommand(cmd, null, false);
            } else if (action.equals(ACTION_KILL)) {
                destroy();
            }
        }
    };
}

File: composeApp/src/androidMain/kotlin/com/hereliesaz/hg2gui/MainManager.kt
package com.hereliesaz.hg2gui

import android.app.Activity
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.graphics.Color
import android.os.Parcelable
import android.text.SpannableString
import android.text.Spanned
import android.text.style.ForegroundColorSpan
import androidx.localbroadcastmanager.content.LocalBroadcastManager
import com.hereliesaz.hg2gui.commands.*
import com.hereliesaz.hg2gui.commands.main.MainPack
import com.hereliesaz.hg2gui.commands.main.raw.location
import com.hereliesaz.hg2gui.commands.main.specific.RedirectCommand
import com.hereliesaz.hg2gui.managers.*
import com.hereliesaz.hg2gui.managers.music.MusicManager2
import com.hereliesaz.hg2gui.managers.music.MusicService
import com.hereliesaz.hg2gui.managers.notifications.KeeperService
import com.hereliesaz.hg2gui.managers.xml.XMLPrefsManager
import com.hereliesaz.hg2gui.managers.xml.options.Behavior
import com.hereliesaz.hg2gui.managers.xml.options.Theme
import com.hereliesaz.hg2gui.tuils.PrivateIOReceiver
import com.hereliesaz.hg2gui.tuils.StoppableThread
import com.hereliesaz.hg2gui.tuils.Tuils
import com.hereliesaz.hg2gui.tuils.interfaces.CommandExecuter
import com.hereliesaz.hg2gui.tuils.interfaces.OnRedirectionListener
import com.hereliesaz.hg2gui.tuils.interfaces.Redirectator
import com.hereliesaz.hg2gui.tuils.interfaces.Reloadable
import com.hereliesaz.hg2gui.tuils.libsuperuser.Shell
import com.hereliesaz.hg2gui.tuils.libsuperuser.ShellHolder
import kotlinx.coroutines.runBlocking
import okhttp3.Cache
import okhttp3.OkHttpClient
import java.io.File
import java.util.regex.Matcher
import java.util.regex.Pattern

/*Copyright Francesco Andreuzzi

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.*/

/**
 * Core logic coordinator for the application.
 */
class MainManager(private val mContext: Activity, reloadable: Reloadable) {

    companion object {
        // Action strings for Intents used in local broadcasts
        const val ACTION_EXEC = "com.hereliesaz.hg2gui.main_exec"
        const val CMD = "cmd"
        const val NEED_WRITE_INPUT = "writeInput"
        const val ALIAS_NAME = "aliasName"
        const val PARCELABLE = "parcelable"
        const val CMD_COUNT = "cmdCount"
        const val MUSIC_SERVICE = "musicService"

        // Static interactive shell session (shared across the app)
        @JvmField
        var interactive: Shell.Interactive? = null

        // Counter to keep track of command order and avoid race conditions
        @JvmField
        var commandCount = 0
    }

    // --- Redirection Logic ---
    private var redirect: RedirectCommand? = null
    private val redirectator = object : Redirectator {
        override fun prepareRedirection(cmd: Any?) {
            val redirectCmd = cmd as? RedirectCommand ?: return
            redirect = redirectCmd
            redirectionListener?.onRedirectionRequest(redirectCmd)
        }

        override fun cleanup() {
            redirect?.let {
                it.beforeObjects.clear()
                it.afterObjects.clear()
                redirectionListener?.onRedirectionEnd(it)
                redirect = null
            }
        }
    }
    
    private var redirectionListener: OnRedirectionListener? = null
    fun setRedirectionListener(redirectionListener: OnRedirectionListener) {
        this.redirectionListener = redirectionListener
    }

    // Package path where raw command classes are located.
    private val COMMANDS_PKG = "com.hereliesaz.hg2gui.commands.main.raw"

    // --- Triggers ---
    private val triggers = arrayOf<CmdTrigger>(
        GroupTrigger(),
        AliasTrigger(),
        TuiCommandTrigger(),
        AppTrigger(),
        ShellCommandTrigger()
    )

    fun interface CommandCompletionListener {
        fun onCommandComplete()
    }
    
    @Volatile
    private var commandCompletionListener: CommandCompletionListener? = null
    fun setCommandCompletionListener(listener: CommandCompletionListener?) {
        commandCompletionListener = listener
    }

    // MainPack holds references to all managers, passed to commands so they can access system resources.
    val mainPack: MainPack

    // Preferences cached for performance
    private val showAliasValue: Boolean
    private val showAppHistory: Boolean
    private val aliasContentColor: Int
    private val multipleCmdSeparator: String
    private val keeperServiceRunning: Boolean

    // Sub-Managers
    private val aliasManager: AliasManager
    private val rssManager: RssManager
    private val appsManager: AppsManager
    private val contactManager: ContactManager?
    private val musicManager2: MusicManager2?
    private val themeManager: ThemeManager
    private val htmlExtractManager: HTMLExtractManager
    private val commandRepository: CommandRepository
    private var messagesManager: MessagesManager? = null

    private val receiver: BroadcastReceiver

    init {
        // Load preferences
        keeperServiceRunning = XMLPrefsManager.getBoolean(Behavior.tui_notification)
        showAliasValue = XMLPrefsManager.getBoolean(Behavior.show_alias_content)
        showAppHistory = XMLPrefsManager.getBoolean(Behavior.show_launch_history)
        aliasContentColor = XMLPrefsManager.getColor(Theme.alias_content_color)
        multipleCmdSeparator = XMLPrefsManager.get(Behavior.multiple_cmd_separator) ?: ""

        // CommandGroup manages categorization of commands
        val group = CommandGroup(mContext, COMMANDS_PKG)

        contactManager = try {
            ContactManager(mContext)
        } catch (e: Exception) {
            Tuils.log(e)
            null
        }

        appsManager = AppsManager(mContext)
        aliasManager = AliasManager(mContext)

        // HTTP Client for network operations (Weather, RSS)
        val client = OkHttpClient.Builder()
            .cache(Cache(mContext.cacheDir, (10 * 1024 * 1024).toLong()))
            .build()

        // Initialize other managers
        rssManager = RssManager(mContext, client)
        themeManager = ThemeManager(client, mContext, reloadable)
        musicManager2 = if (XMLPrefsManager.getBoolean(Behavior.enable_music)) MusicManager2(mContext) else null
        ChangelogManager.printLog(mContext, client)
        htmlExtractManager = HTMLExtractManager(mContext, client)

        if (XMLPrefsManager.getBoolean(Behavior.show_hints)) {
            messagesManager = MessagesManager(mContext)
        }

        // Initialize Command Repository (indexes available commands)
        commandRepository = CommandRepository()
        // Create the MainPack data transfer object
        mainPack = MainPack(
            mContext, group, aliasManager, appsManager, musicManager2,
            contactManager, redirectator, rssManager, client, commandRepository
        )
        // Populate command repository with available commands
        commandRepository.update(mainPack)

        // Initialize Shell
        val shellHolder = ShellHolder(mContext)
        interactive = shellHolder.build()
        mainPack.shellHolder = shellHolder

        // Register BroadcastReceiver for internal events
        val filter = IntentFilter().apply {
            addAction(ACTION_EXEC)
            addAction(location.ACTION_LOCATION_CMD_GOT)
            addAction(UIManager.ACTION_UPDATE_SUGGESTIONS)
        }

        receiver = object : BroadcastReceiver() {
            override fun onReceive(context: Context, intent: Intent) {
                val action = intent.action
                if (action == UIManager.ACTION_UPDATE_SUGGESTIONS) {
                    commandRepository.update(mainPack)
                } else if (action == ACTION_EXEC) {
                    var cmd = intent.getStringExtra(CMD) ?: intent.getStringExtra(PrivateIOReceiver.TEXT) ?: return

                    // Check for stale commands
                    val cmdCount = intent.getIntExtra(CMD_COUNT, -1)
                    if (cmdCount < commandCount) return
                    commandCount++

                    val aliasName = intent.getStringExtra(ALIAS_NAME)
                    val needWriteInput = intent.getBooleanExtra(NEED_WRITE_INPUT, false)
                    val p = intent.getParcelableExtra<Parcelable>(PARCELABLE)

                    // If requested, echo the command to the input field
                    if (needWriteInput) {
                        val i = Intent(PrivateIOReceiver.ACTION_INPUT).apply {
                            putExtra(PrivateIOReceiver.TEXT, cmd)
                        }
                        LocalBroadcastManager.getInstance(context.applicationContext).sendBroadcast(i)
                    }

                    // Execute based on type
                    if (p != null && p is AppsManager.LaunchInfo) {
                        onCommand(cmd, p, intent.getBooleanExtra(MUSIC_SERVICE, false))
                    } else {
                        onCommand(cmd, aliasName, intent.getBooleanExtra(MUSIC_SERVICE, false))
                    }
                } else if (action == location.ACTION_LOCATION_CMD_GOT) {
                    // Handle async location result
                    Tuils.sendOutput(
                        context,
                        "Lat: ${intent.getDoubleExtra(TuiLocationManager.LATITUDE, 0.0)}; Long: ${
                            intent.getDoubleExtra(TuiLocationManager.LONGITUDE, 0.0)
                        }"
                    )
                    TuiLocationManager.instance(context).rm(location.ACTION_LOCATION_CMD_GOT)
                }
            }
        }

        LocalBroadcastManager.getInstance(mContext.applicationContext).registerReceiver(receiver, filter)
    }

    /**
     * Updates background services when a command is executed.
     */
    private fun updateServices(cmd: String, wasMusicService: Boolean) {
        if (keeperServiceRunning) {
            val i = Intent(mContext, KeeperService::class.java).apply {
                putExtra(KeeperService.CMD_KEY, cmd)
                putExtra(KeeperService.PATH_KEY, mainPack.currentDirectory?.absolutePath)
            }
            mContext.startService(i)
        }

        if (wasMusicService) {
            val i = Intent(mContext, MusicService::class.java)
            mContext.startService(i)
        }
    }

    fun onCommand(input: String, launchInfo: AppsManager.LaunchInfo?, wasMusicService: Boolean) {
        if (launchInfo == null) {
            onCommand(input, null as String?, wasMusicService)
            return
        }

        updateServices(input, wasMusicService)

        // Verify if the input matches the app label
        if (launchInfo.unspacedLowercaseLabel == Tuils.removeSpaces(input.lowercase())) {
            performLaunch(mainPack, launchInfo, input)
        } else {
            // Fallback to standard processing
            onCommand(input, null as String?, wasMusicService)
        }
    }

    private val colorExtractor = Pattern.compile("(#[^(]{6})\\[([^\\)]*)\\]", Pattern.CASE_INSENSITIVE)

    fun onCommand(input: String, alias: String?, wasMusicService: Boolean) {
        var processedInput = Tuils.removeUnncesarySpaces(input)

        if (alias == null) updateServices(processedInput, wasMusicService)

        // --- Redirection Handling ---
        redirect?.let {
            if (!it.isWaitingPermission) {
                it.afterObjects.add(processedInput)
            }
            val output = it.onRedirect(mainPack)
            Tuils.sendOutput(mContext, output)
            return
        }

        // Show alias expansion if enabled
        if (alias != null && showAliasValue) {
            Tuils.sendOutput(aliasContentColor, mContext, aliasManager.formatLabel(alias, processedInput))
        }

        // --- Multiple Commands ---
        val cmds = if (multipleCmdSeparator.isNotEmpty()) {
            processedInput.split(multipleCmdSeparator.toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
        } else {
            arrayOf(processedInput)
        }

        // --- Execution Loop ---
        for (c in cmds.indices) {
            var cmd = cmds[c]
            var color = TerminalManager.NO_COLOR

            val m = colorExtractor.matcher(cmd)
            if (m.matches()) {
                try {
                    color = Color.parseColor(m.group(1))
                    cmd = m.group(2)
                } catch (e: Exception) {
                    color = TerminalManager.NO_COLOR
                }
            }

            mainPack.clear()
            mainPack.commandColor = color

            // Iterate through triggers until one handles the command
            for (trigger in triggers) {
                val r = try {
                    trigger.trigger(mainPack, cmd)
                } catch (e: Exception) {
                    Tuils.sendOutput(mContext, Tuils.getStackTrace(e))
                    break
                }
                if (r) {
                    messagesManager?.afterCmd()
                    break
                }
            }
        }
    }

    fun onLongBack() {
        Tuils.sendInput(mContext, Tuils.EMPTYSTRING)
    }

    fun sendPermissionNotGrantedWarning() {
        redirectator.cleanup()
    }

    fun dispose() {
        mainPack.dispose()
    }

    fun destroy() {
        mainPack.destroy()
        TuiLocationManager.disposeStatic()
        messagesManager?.onDestroy()
        themeManager.dispose()
        htmlExtractManager.dispose(mContext)
        aliasManager.dispose()
        LocalBroadcastManager.getInstance(mContext.applicationContext).unregisterReceiver(receiver)

        object : StoppableThread() {
            override fun run() {
                super.run()
                try {
                    interactive?.kill()
                    interactive?.close()
                } catch (e: Exception) {
                    Tuils.log(e)
                    Tuils.toFile(e)
                }
            }
        }.start()
    }

    fun executer(): CommandExecuter {
        return CommandExecuter { input, obj ->
            val li = if (obj is AppsManager.LaunchInfo) obj else null
            onCommand(input, li, false)
        }
    }

    private var appFormat: String? = null
    private var outputColor: Int = 0

    private val pa = Pattern.compile("%a", Pattern.CASE_INSENSITIVE or Pattern.LITERAL)
    private val pp = Pattern.compile("%p", Pattern.CASE_INSENSITIVE or Pattern.LITERAL)
    private val pl = Pattern.compile("%l", Pattern.CASE_INSENSITIVE or Pattern.LITERAL)

    fun performLaunch(mainPack: MainPack, i: AppsManager.LaunchInfo, input: String): Boolean {
        val intent = appsManager.getIntent(i) ?: return false

        // Show launch history message if enabled
        if (showAppHistory) {
            if (appFormat == null) {
                appFormat = XMLPrefsManager.get(Behavior.app_launch_format)
                outputColor = XMLPrefsManager.getColor(Theme.output_color)
            }

            var a = appFormat!!
            a = pa.matcher(a).replaceAll(Matcher.quoteReplacement(intent.component!!.className))
            a = pp.matcher(a).replaceAll(Matcher.quoteReplacement(intent.component!!.packageName))
            a = pl.matcher(a).replaceAll(Matcher.quoteReplacement(i.publicLabel))
            a = Tuils.patternNewline.matcher(a).replaceAll(Matcher.quoteReplacement(Tuils.NEWLINE))

            val text = SpannableString(a).apply {
                setSpan(ForegroundColorSpan(outputColor), 0, length, Spanned.SPAN_EXCLUSIVE_EXCLUSIVE)
            }
            val s = TimeManager.instance?.replace(text) ?: text

            Tuils.sendOutput(mainPack, s, TerminalManager.CATEGORY_OUTPUT)
        }

        Tuils.getContext(mainPack).startActivity(intent)
        return true
    }

    interface CmdTrigger {
        @Throws(Exception::class)
        fun trigger(info: MainPack, input: String): Boolean
    }

    private inner class AliasTrigger : CmdTrigger {
        override fun trigger(info: MainPack, input: String): Boolean {
            val alias = aliasManager.getAlias(input, true)
            val aliasValue = alias[0] ?: return false
            val aliasName = alias[1]
            val residual = alias[2]

            val expandedValue = aliasManager.format(aliasValue, residual ?: "")
            onCommand(expandedValue, aliasName, false)
            return true
        }
    }

    private inner class GroupTrigger : CmdTrigger {
        override fun trigger(info: MainPack, input: String): Boolean {
            val spaceIndex = input.indexOf(Tuils.SPACE)
            val name: String
            val remainingInput: String?

            if (spaceIndex != -1) {
                name = input.substring(0, spaceIndex)
                remainingInput = input.substring(spaceIndex + 1)
            } else {
                name = input
                remainingInput = null
            }

            val appGroups = info.appsManager.groups
            if (appGroups != null) {
                for (g in appGroups) {
                    if (name == g.name()) {
                        return if (remainingInput == null) {
                            @Suppress("UNCHECKED_CAST")
                            Tuils.sendOutput(
                                mContext,
                                AppsManager.AppUtils.printApps(
                                    AppsManager.AppUtils.labelList(g.members() as List<AppsManager.LaunchInfo>, false)
                                )
                            )
                            true
                        } else {
                            g.use(mainPack, remainingInput)
                        }
                    }
                }
            }
            return false
        }
    }

    private inner class ShellCommandTrigger : CmdTrigger {
        private val CD_CODE = 10
        private val PWD_CODE = 11

        private val result: Shell.OnCommandResultListener = object : Shell.OnCommandResultListener {
            override fun onCommandResult(commandCode: Int, exitCode: Int, output: MutableList<String>) {
                if (commandCode == CD_CODE) {
                    interactive?.addCommand("pwd", PWD_CODE, this)
                } else if (commandCode == PWD_CODE && output.size == 1) {
                    val f = File(output[0])
                    if (f.exists()) {
                        mainPack.currentDirectory = f
                        LocalBroadcastManager.getInstance(mContext.applicationContext)
                            .sendBroadcast(Intent(UIManager.ACTION_UPDATE_HINT))
                    }
                }
            }
        }

        override fun trigger(info: MainPack, input: String): Boolean {
            object : StoppableThread() {
                override fun run() {
                    if (input.trim().equals("su", ignoreCase = true)) {
                        if (Shell.SU.available()) {
                            LocalBroadcastManager.getInstance(mContext.applicationContext)
                                .sendBroadcast(Intent(UIManager.ACTION_ROOT))
                        }
                        interactive?.addCommand("su")
                    } else if (input.contains("cd ")) {
                        interactive?.addCommand(input, CD_CODE, result)
                    } else {
                        interactive?.addCommand(input)
                    }
                }
            }.start()
            return true
        }
    }

    private inner class AppTrigger : CmdTrigger {
        override fun trigger(info: MainPack, input: String): Boolean {
            val i = appsManager.findLaunchInfoWithLabel(input, AppsManager.SHOWN_APPS)
            return i != null && performLaunch(info, i, input)
        }
    }

    private inner class TuiCommandTrigger : CmdTrigger {
        override fun trigger(info: MainPack, input: String): Boolean {
            val command = CommandTuils.parse(input, info)
            if (command == null) {
                commandCompletionListener?.onCommandComplete()
                return false
            }

            mainPack.lastCommand = input

            object : StoppableThread() {
                override fun run() {
                    super.run()
                    try {
                        // execStream() was declared on StreamableCommand but nothing ever called
                        // it - this always fell through to exec() below instead, so a
                        // StreamableCommand's real work (e.g. bootstrap's download) never
                        // happened; its exec() only ever returns a placeholder string.
                        val stream = command.execStream(info)
                        if (stream != null) {
                            runBlocking {
                                stream.collect { line ->
                                    Tuils.sendOutput(info, line, TerminalManager.CATEGORY_OUTPUT)
                                }
                            }
                        } else {
                            val output = command.exec(info)
                            if (output != null) {
                                Tuils.sendOutput(info, output, TerminalManager.CATEGORY_OUTPUT)
                            }
                        }
                    } catch (e: Exception) {
                        Tuils.sendOutput(mContext, Tuils.getStackTrace(e))
                        Tuils.log(e)
                    } finally {
                        commandCompletionListener?.onCommandComplete()
                    }
                }
            }.start()
            return true
        }
    }

    interface Group {
        fun members(): List<*>?
        fun use(mainPack: MainPack, input: String): Boolean
        fun name(): String?
    }
}
